docs/proposals/SymbolicShapeInfProposal.md

```diff
@@ -0,0 +1,163 @@
+
+Note: This work will benefit Nuphar as well but right now there is no plan to move Nuphar to use this solution.  
+  
+
+## Terminology 
+Shape inference can be broken into 2 parts:  
+
+* Node level shape inference: This refers to operator specific shape inference functions. They are defined with the operator schema itself.  
+
+* Graph-level shape inference: This refers to the higher-level logic which walks through the entire graph, gets the inferred shape from node level shape inference functions and then makes decisions on merging these inferred shapes with existing shapes so that they are available for downstream nodes.  
+ 
+
+## Proposal 
+Extend current shape inference to allow:  
+* Symbol generation and propagation 
+
+* Partial data computation and propagation 
+
+* Extend shape op to generate slice of the shape to facilitate simplifying shape computations.  
+
+
+## Extend shape inference  
+
+### Symbol generation and propagation 
+Extend graph level shape inference to maintain a graph level view of symbols and generate new symbols where necessary. This will enable us to continue the shape inference of the downstream nodes. 
+
+Example: 
+
+For an op like “Concat” if its inputs have shapes “[M]” and “[N]” current shape-inference returns “[?]” where “?” is to indicate a dimension with neither dim-value nor dim-param set. Now, suppose the output X of “Concat” is input to a unary-op Op1() whose output Y is then input to another unary-op Op2() whose output is Z, etc. The shape “[?]” is propagated further. We infer that Y and Z have shape “[?]”. However, we do not infer that X, Y, and Z have the same shape because two “?” cannot be considered equal.  
+
+Per the current proposal, “[?]” in inferred shapes will be replaced by a new unique symbol by the graph level shape inference so the downstream nodes can use the symbolic shapes to carry out shape inference. In the current example, “Concat” will produce “[?]” as the shape which will then be replaced by “[K]”, then subsequent shape inference will infer that X, Y, and Z all have the same shape “[K]”. Runtimes can use this information to reuse memory for these tensors.  
+
+
+### Partial data computation and propagation 
+When shape inputs are computed dynamically, shape inference post a reshape node stops. This can be prevented by making this data available to the reshape node during shape inference. We propose computation and propagation of data for operators which are used in shape computation. 
+
+It is called “partial” data computation and propagation because this will only be done for shape computations. It is not meant to be a full-fledged kernel for the operator. For the same reasons data computations will be implemented for a limited set of operators. While we will increase the coverage in the future iterations it is important to note that for some operators like LSTM, convolution ops, pooling ops etc. this will never be added because such ops are not used in shape computations. 
```

Comment:
```suggestion
It is called “partial” data computation and propagation because this will only be done for shape computations. It is not meant to be a full-fledged kernel for the operator. For the same reasons data computations will be implemented for a limited set of operators. While we will increase the coverage in the future iterations it is important to note that for some operators like LSTM, convolution ops, pooling ops etc. These will never be added because such ops are not used in shape computations. 
```

Reply:
this refers to data propagation function. Updated to use "data propagation function"